release/nightly_tests/dataset/map_batches_benchmark.py

```diff
@@ -0,0 +1,128 @@
+        "s3://shuffling-data-loader-benchmarks/data/input_data_0.parquet.snappy"
+    )
+    lazy_input_ds = input_ds.lazy()
+
+    batch_formats = ["pandas", "numpy"]
+    batch_sizes = [64, 128, 256, 512, 1024, 2048, 4096, None]
+    num_calls_list = [1, 2, 4, 8, 16]
+
+    # Test different batch_size of map_batches.
+    for batch_format in batch_formats:
+        for batch_size in batch_sizes:
+            # TODO(chengsu): Investigate why NumPy with batch_size less than 512,
+            # took forever to finish.
+            if batch_format == "numpy" and batch_size is not None and batch_size < 512:
+                continue
+
+            num_calls = 2
+            test_name = f"map-batches-{batch_format}-{batch_size}-{num_calls}-default"
+            benchmark.run(
+                test_name,
+                map_batches,
+                input_ds=input_ds,
+                batch_format=batch_format,
+                batch_size=batch_size,
+                num_calls=num_calls,
+            )
+            test_name = f"map-batches-{batch_format}-{batch_size}-{num_calls}-lazy"
+            benchmark.run(
+                test_name,
+                map_batches,
+                input_ds=lazy_input_ds,
+                batch_format=batch_format,
+                batch_size=batch_size,
+                num_calls=num_calls,
+            )
+
+    # Test multiple calls of map_batches.
+    for batch_format in batch_formats:
+        for num_calls in num_calls_list:
+            for compute in ["tasks", ActorPoolStrategy(1, 1)]:
+                batch_size = 4096
+                if compute == "tasks":
+                    compute_strategy = "task"
+                else:
+                    compute_strategy = "actor"
+
+                test_name = (
+                    f"map-batches-{batch_format}-{batch_size}-{num_calls}-"
+                    f"{compute_strategy}-default"
+                )
+                benchmark.run(
+                    test_name,
+                    map_batches,
+                    input_ds=input_ds,
+                    batch_format=batch_format,
+                    batch_size=batch_size,
+                    compute=compute,
+                    num_calls=num_calls,
+                )
+                test_name = (
+                    f"map-batches-{batch_format}-{batch_size}-{num_calls}-"
+                    f"{compute_strategy}-lazy"
+                )
+                benchmark.run(
+                    test_name,
+                    map_batches,
+                    input_ds=lazy_input_ds,
+                    batch_format=batch_format,
+                    batch_size=batch_size,
+                    compute=compute,
+                    num_calls=num_calls,
+                )
+
+    # Test different batch formats of map_batches.
+    for current_format in ["pyarrow", "pandas", "numpy"]:
+        new_input_ds = input_ds.map_batches(
+            lambda ds: ds, batch_format=current_format, batch_size=None
```

Comment:
The eventual conversion between Arrow tables and NumPy ndarrays becomes much more expensive when the upstream Arrow tables are chunked, e.g. due to an upstream `.map_batches(..., batch_size=chunk_size)` or due to batched/buffered reading, since the ndarrays will need to be contiguous in memory, which will require the chunks to be consolidated. We even have an open performance issue for this: https://github.com/ray-project/ray/issues/28526

It might be useful to capture upstream chunking/concatenation in this benchmark.

Reply:
I agree, added test for different batch sizes.